craft_parts/permissions.py

```diff
@@ -48,11 +48,12 @@ class Permissions(BaseModel):
     mode: Optional[str] = None
 
     # pylint: disable=no-self-argument
-    @root_validator(pre=True)
-    def validate_root(cls, values: Dict[Any, Any]) -> Dict[Any, Any]:
+    @model_validator(mode="after")
```

Comment:
same here `mode="before"`

Reply:
> what's the need to go from pyyaml to ruamel?
According to @cmatsuoka, pydantic2 likes ruamel better and some tests will need adjustments if we need to keep using pyyaml. ruamel on the other hand makes the transition to pydantic2 smoother, and since ruamel supports YAML 1.2 format and also preserves the order of dictionaries and the formatting of the original YAML file: that's an additional advantage. Virtually, switching to ruamel should pose no issue, but @mr-cal raised some concerns about it in the comments above.

Reply:
> specifically, this would break because source-tag is a float and not a string right? I think this is "too breaking" of a change: it would break all snaps, rocks, charms etc that use things that look like numbers.
This is also mostly unrelated to migrating to v2: if possible, we should stick to the minimum required to do the update with as few changes as possible.

source-tag is a string by definition in Parts class, and in our units tests we only test tags like "v2.3". Is it a common practice for people writing *craft.yaml files to use unquoted numeric values there?
I carried over the "strict" field from Claudio's experimental branch, but if you think that some fields should not be strict-typed due to common practice or other reasons then it should be removed.

Reply:
I don't know if it's a common practice but I would expect many cases out there to use things like `1.0` for the tag. And source-tag is just one of the properties, I imagine that many things might break with the other properties if we enable this strictness

Reply:
> shouldn't this be mode="before", to keep the semantics of pre=True?

You would assume `mode="before"` is just a replacement for `pre=True`, and the documention will only help you believe so! But try to replace that one word "after" with "before" and most unit tests will go red:
![image](https://github.com/canonical/craft-parts/assets/719022/cf9cea99-03a0-4b35-b668-20339e56dbb8)

Apparently, you need to to the model validation after the field validation, because otherwise the model will throw an error because it can't find the fields that weren't validated yet!